### ISSUE TITLE
Cosine Annealing

### ISSUE THREAD

**Issue by NicoRenaud** [DeepRank/deeprank#30]
Should we use cosine annealing or early stopping ?

https://www.reddit.com/r/MachineLearning/comments/6h1kry/d_heuristics_for_early_stopping/
https://discuss.pytorch.org/t/current-learning-rate-and-cosine-annealing/8952
https://openreview.net/pdf?id=BJYwwY9ll
https://arxiv.org/pdf/1608.03983.pdf

**Comment by NicoRenaud** [comment]
Cosine annealing is implemented in the master branch of PyTorch 
http://pytorch.org/docs/master/optim.html
But we are using the stable release ....

**Comment by DCLukas** [comment]
One additional option when using cosine annealing is the use of snapshot ensembles.
https://arxiv.org/abs/1704.00109
After each cosine cycle save the current model and ensemble these for a more generalizable model. Also related to issue #24

**Comment by DCLukas** [comment]
In addition, the latest stable release (4.0) does provide Cosine annealing:

`scheduler = optim.lr_scheduler.CosineAnnealingLR($optimizer, $numOfBatchesPerEpoch)`

Then calling _scheduler.step()_ after _loss.backward()_ should do the trick.

**Comment by sonjageorgievska** [comment]
I will close this for now as it is not critical and we want a cleaner issue list and focus before release
